proposed/2022/DotnetUpdatePackage.md

```diff
@@ -0,0 +1,179 @@
+# Title
+
+- [Jon Douglas](https://github.com/JonDouglas)
+- Start Date (2022-05-11)
+- [#4103](https://github.com/NuGet/Home/issues/4103)
+
+## Summary
+
+<!-- One-paragraph description of the proposal. -->
+Staying current on the latest package version is a major challenge for modern .NET developers. Developers would like an update all experience in the .NET CLI similar to what they can do in modern IDEs like Visual Studio and checking all packages to then be updated in bulk. This feature will support the bulk updating of eligible `<PackageReference>` elements throughout their project files and solutions.
+
+`dotnet update package` can automatically update any outdated package(s) to the latest version depending on the update strategy provided by the user.
+
+## Motivation
+
+<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
+To rid developers of "DLL Hell" in managing modern .NET dependencies, we would like to implement this feature to rid developers of daily pain of managing their dependencies and staying current with the latest versions of packages to maintain a secure supply chain and new package feature functionality.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+This command will update dependencies to the latest version respecting the semver constraints of your package and its transitive dependencies. It will also install any missing packages if an newer version brings them in. It will take into account lock files and central package management as locations that exist in addition to common definitions of `<PackageReference>` in project file(s).
+
+To ensure compatibility with existing package definitions in various files, we should not overwrite or simplify XML elements, but rather only update package version attributes.
+
+![](../../meta/resources/dotnetupdatepackage/dotnetupdatepackage.png)
+
+```
+dotnet update [<PROJECT>|<SOLUTION>] package <PACKAGE_NAME>
+    [-f|--framework <FRAMEWORK>] [--interactive]
+    [-n|--no-restore] [--package-directory <PACKAGE_DIRECTORY>]
+    [--prerelease] [-s|--source <SOURCE>] [-v|--version <VERSION>]
+    [--dry-run] [--highest-minor] [--highest-patch]
+
+dotnet update package -h|--help
+```
+
+#### Arguments
+
+- PROJECT | SOLUTION
```

Comment:
Is there way to specify a directory of projects? i.e. I want to update all my packages in my \test directory?

Reply:
We'll need to limit the scope. So let's start with project or solution. Updating the world would be amazing, but very complex I think!

Reply:
Is it? This design already necessitates the ability to "find a project" within the current directory when the <PROJECT> isn't specific. Might be oversimplifying, but isn't the functionality I'm describing just a loop to do that for all projects recursively within a directory?

Reply:
I believe it's possible. For example if you ran the command on the repo folder, it should work recursively.

Reply:
Based on this line:

> The project or solution file to operate on. If not specified, the command searches the current directory for one. If more than one solution or project is found, an error is thrown.

The current design will error out if you run it on the repo folder because multiple projects will be found. That's why I'm double checking. My expectation would be that it would run recursively/ apply to all found projects.

Reply:
Let's change this language then!

Reply:
How many folders deep would you search for projects? 
All of it? 

I think deciding the `target` of a command should be something we involve the SDK team on given that `package` is a first class noun in dotnet.exe.

Reply:
@baronfel Any idea how we can get more eyes from SDK team on these types of things? :)

Reply:
You can unfortunately you can't tag @dotnet/dotnet-cli, so I'll poke a few here: @dsplaisted, @joeloff, @gkulin

I'd generally be in favor of globs/patterns for more SDK operations - I know a similar capability has been requested for many of the MSBuild-driving commands in the past.  It's common to use globs for this kind of thing - `**/*Test.*proj` to grab every test project in a solution recursively, for example.